File: src/c4/inter/Unary.java
package inter;

import lexer.*;
import code.DreiAdrCode;
import code.Arith1OpCode;
import treewalker.TreeWalker;

/*
 * Unary ist eine Unterklasse von Op und beschreibt unäre arithmetische Ausdrücke. 
 * In der Instanzenvariablen expr ist der Operand abgelegt
 */

public class Unary extends Op {
	Expr expr;

	public Unary(Token tok, Expr x) { // behandelt unäres minus, für ! siehe Not
		super(tok);
		expr = x;
	}

	public Expr getExpr() {
		return expr;
	}

	public void setExpr(Expr expr) {
		this.expr = expr;
	}


	// für die Drei-Adress-Code Erzeugung
	public DreiAdrCode codeForValueTo(Id id) {
		return (new Arith1OpCode(id, this));
	}

	public <R, P> R walk(TreeWalker<R, P> walker, P arg) {
		return walker.walkUnaryNode(this, arg);
	}

}

File: src/c4/inter/Unary.rs
use super::{
    DreiAdrCode, Expr, ExprData, ExpressionNode, Id, Node, NodeData, OpData, Token, TreeWalker,
    Walk,
};

/// Unary arithmetic expression. Logical negation is represented by `Not`.
#[derive(Clone, Debug, PartialEq)]
pub struct Unary {
    op: OpData,
    expr: Expr,
}

impl Unary {
    pub fn new(token: Token, expr: Expr) -> Self {
        Self {
            op: OpData::new(token),
            expr,
        }
    }

    pub fn get_expr(&self) -> &Expr {
        &self.expr
    }
    pub fn set_expr(&mut self, expr: Expr) {
        self.expr = expr;
    }

    pub fn code_for_value_to(&self, target: Id) -> DreiAdrCode {
        let value = self
            .expr
            .as_singleton()
            .expect("unary operand must be lowered to a Singleton");
        DreiAdrCode::Unary {
            target,
            operator: self.get_op().clone(),
            value,
        }
    }
}

impl Node for Unary {
    fn node_data(&self) -> &NodeData {
        &self.op.expr.node
    }
}

impl ExpressionNode for Unary {
    fn expr_data(&self) -> &ExprData {
        &self.op.expr
    }
    fn expr_data_mut(&mut self) -> &mut ExprData {
        &mut self.op.expr
    }
}

impl<P, R> Walk<P, R> for Unary {
    fn walk<W: TreeWalker<P, R> + ?Sized>(&self, walker: &mut W, arg: P) -> R {
        walker.walk_unary_node(self, arg)
    }
}
